src/Imgix/UrlBuilder.cs

```diff
@@ -56,40 +58,217 @@ public String BuildUrl(String path)
+        * characters, like '/' and '@', alone while encoding the rest using `EncodeURI`.
+        */
+        private String SanitizePath(String p)
+        {
+            String path = p;
+            // Strip leading slash first (we'll re-add after encoding).
+            path = path.TrimEnd('/').TrimStart('/');
 
+            // Check if path is a proxy path and store type of proxy.
+            Dictionary<String, Boolean> proxyStatus = CheckProxyStatus(path);
+            Object pathIsProxy = proxyStatus[IS_PROXY];
+            Object proxyIsEncoded = proxyStatus[IS_ENCODED];
+
+            if (pathIsProxy.Equals(true) && proxyIsEncoded.Equals(false))
+            {
+                // Use WebUtility.UrlEncode to ensure *all* characters are handled,
+                // since it's being used as a path.
+                return WebUtility.UrlEncode(path);
+            }
+            else if (pathIsProxy.Equals(true) && proxyIsEncoded.Equals(true))
+            {
+                // Do nothing to URL being used as path since its encoded.
+                return path;
+            }
+            else
+            {
+                // The path is not being used as a proxy, so leave
+                // legal characters like '/' and '@' alone.
+                return EncodeURI(path);
+            }
+        }
+
+        /*
+        * EncodeURI encodes string `p` to the UTF8 format. It splits
+        * the string on the `/` character to avoid accidentally
+        * encoding them. It then iterates over every character and
+        * calls the `UrlEncode` method on them. Finally, it joins
+        * the string array on `/`, adding the removed character back.
+        */
+        private String EncodeURI(String p)
+        {
+            string[] splitPath = p.Split('/');
+            List<String> sanitizedPath = new List<String>();
+
+            foreach (String s in splitPath)
+            {
+                sanitizedPath.Add(UrlEncode(s));
+            }
+
+            return string.Join("/", sanitizedPath);
+        }
+
+
+        /* 
+        * CheckProxyStatus checks if the path has one of the four possible
+        * acceptable proxy prefixes. First we check if the path has the
+        * correct ascii prefix. If it does then we know that it is a proxy,
+        * but it's not percent encoded. Second, we check if the path is
+        * prefixed by a percent-encoded prefix. If it is, we know that it's
+        * a proxy and that it's percent-encoded. Finally, if the path isn't
+        * prefixed by any of these four prefixes, it is not a valid proxy.
+        * This might be "just enough validation," but if we run into issues
+        * we can make this check smarter/more-robust.
+        */
+        private static Dictionary<String, Boolean> CheckProxyStatus(String p)
+        {
+            String path = p;
+            Dictionary<String, Boolean> status = new Dictionary<String, Boolean>();
+            path.Replace("^/", "");
```

Comment:
out of curiosity, you already trim the leading and trailing slashes in [SanitizePath](https://github.com/imgix/imgix-csharp/pull/40/commits/8068ee09051f2eb0558d3db453630f3b267b140c#diff-7ae8782884590f68d4dc070e0791a2d2d4c30a536f3ed7c2bc46728467b729e7R99) so do we still need to this line?

Reply:
💯  good catch, not needed here.